### ISSUE TITLE
initialScene variable has a wrong type

### ISSUE THREAD

**Issue by vkaramov** [SwiftGen/SwiftGen#448]
Xcode 9.2, OS X 10.13.5, wiftGen v5.3.0 (Stencil v0.10.1, StencilSwiftKit v2.4.0, SwiftGenKit v2.2.0) :
Variable sceneClass is empty for some reason here

> {% set sceneClass %}{% call className storyboard.initialScene %}{% endset %}

so that 

generated code looks like:
```
internal enum StoryboardScene {
  internal enum AccountSettings: StoryboardType {
    internal static let storyboardName = "AccountSettings"

    internal static let initialScene = InitialSceneType<>(storyboard: AccountSettings.self)
```

**Comment by AliSoftware** [comment]
Could you share your `AccountSettings.storyboard` file? (Reducing it down and Removing any confidential design or content beforehand if necessary)?
Thanks

**Comment by vkaramov** [comment]
I've attached both script and template as well as storyboard file which generates the wrong output.
[Script.zip](https://github.com/SwiftGen/SwiftGen/files/2169756/Script.zip)

[Login_iPhone.storyboard.zip](https://github.com/SwiftGen/SwiftGen/files/2169754/Login_iPhone.storyboard.zip)

**Comment by vkaramov** [comment]
Generated file:

```
// Generated using SwiftGen, by O.Halligon — https://github.com/SwiftGen/SwiftGen

// swiftlint:disable sorted_imports
import Foundation

// swiftlint:disable superfluous_disable_command
// swiftlint:disable file_length

internal protocol StoryboardType {
  static var storyboardName: String { get }
}

internal extension StoryboardType {
  static var storyboard: UIStoryboard {
    let name = self.storyboardName
    return UIStoryboard(name: name, bundle: Bundle(for: BundleToken.self))
  }
}

internal struct SceneType<T: Any> {
  internal let storyboard: StoryboardType.Type
  internal let identifier: String

  internal func instantiate() -> T {
    let identifier = self.identifier
    guard let controller = storyboard.storyboard.instantiateViewController(withIdentifier: identifier) as? T else {
      fatalError("ViewController '\(identifier)' is not of the expected class \(T.self).")
    }
    return controller
  }
}

internal struct InitialSceneType<T: Any> {
  internal let storyboard: StoryboardType.Type

  internal func instantiate() -> T {
    guard let controller = storyboard.storyboard.instantiateInitialViewController() as? T else {
      fatalError("ViewController is not of the expected class \(T.self).")
    }
    return controller
  }
}

internal protocol SegueType: RawRepresentable { }

internal extension UIViewController {
  func perform<S: SegueType>(segue: S, sender: Any? = nil) where S.RawValue == String {
    let identifier = segue.rawValue
    performSegue(withIdentifier: identifier, sender: sender)
  }
}

// swiftlint:disable explicit_type_interface identifier_name line_length type_body_length type_name
internal enum StoryboardScene {
  internal enum LoginIPhone: StoryboardType {
    internal static let storyboardName = "Login_iPhone"

    internal static let initialScene = InitialSceneType<>(storyboard: LoginIPhone.self)

    internal static let loginViewController = SceneType<>(storyboard: LoginIPhone.self, identifier: "LoginViewController")

    internal static let termsViewController = SceneType<>(storyboard: LoginIPhone.self, identifier: "TermsViewController")
  }
}

internal enum StoryboardSegue {
}
// swiftlint:enable explicit_type_interface identifier_name line_length type_body_length type_name

private final class BundleToken {}
```

**Comment by vkaramov** [comment]
@AliSoftware Just pulled the latest changes from https://github.com/AliSoftware/SwiftGen.git and build with Xcode 9.4.1. Works better:
```
internal enum StoryboardScene {
  internal enum AccountSettings: StoryboardType {
    internal static let storyboardName = "AccountSettings"

    internal static let initialScene = InitialSceneType<MyCoolModule.AccountSettingsTVC>(storyboard: AccountSettings.self)

    internal static let mySettings = SceneType<MyCoolModule.AccountSettingsTVC>(storyboard: AccountSettings.self, identifier: "my-settings")
  }
```

And it doesn't compile because of the module name:

> 'AccountSettingsTVC' is not a member type of 'StoryboardScene.MyCoolModule'

**Comment by vkaramov** [comment]
@AliSoftware I found a workaround here for the **current** development **snapshot** - 
I just replaced sceneClass definition:
Instead of:
`    {% set sceneClass %}{% call className storyboard.initialScene %}{% endset %}
    {{accessModifier}} static let initialScene = InitialSceneType<{{sceneClass}}>(storyboard: {{storyboardName}}.self)
`
I wrote 
`    {% set sceneClass %}{{storyboard.initialScene.type}}{% endset %}
    {{accessModifier}} static let initialScene = InitialSceneType<{{sceneClass}}>(storyboard: {{storyboardName}}.self)
`

and the similar approach below:
`    {% set sceneClass %}{{scene.type}}{% endset %}
    {{accessModifier}} static let {{sceneID}} = SceneType<{{sceneClass}}>(storyboard: {{storyboardName}}.self, identifier: "{{scene.identifier}}")
`

Could you please publish the latest version of SwiftGen? I don't really like to push SwiftGen binary to our Git repository.

**Comment by djbe** [comment]
Hi @vkaramov,

Your quick fix removes the whole module part of that macro call, which would break the template for most users.

We're going to need a bit of time to check why it isn't working in swiftgen 5.3, and if we need to fix anything in our codebase. It's weird that the current `master` does work for you, as I can't remember changing anything about the storyboard parser except for #367, which shouldn't affect you (as you'd have an empty result file in that case).

**Comment by vkaramov** [comment]
@djbe It also doesn't work for 5.2 and 5.1 as well.

**Comment by vkaramov** [comment]
@djbe As for the #367, I saw that issue with 5.3 when tried to sun it for another iOS project.

**Comment by AliSoftware** [comment]
What's strange is that the `.storyboard` files properly contains the `customClass` attribute in its XML:

```xml
<viewController storyboardIdentifier="LoginViewController"
  definesPresentationContext="YES"
  modalTransitionStyle="flipHorizontal"
  useStoryboardIdentifierAsRestorationIdentifier="YES"
  id="eAB-O9-dR8"
  customClass="LoginViewController"
  sceneMemberID="viewController">
…
```

So it shouldn't depend on the version of Xcode, and it also looks similar to other Storyboards we have in our unit tests, so I wonder what's different…

We'll indeed need some time to investigate, we'll keep you posted.

**Comment by vkaramov** [comment]
@AliSoftware  @djbe  when do expect to issue the new public version?

**Comment by AliSoftware** [comment]
Honestly no idea. We hope soon, but will depend on our available time (personally no time at all during work ours lately, my paid work is pretty full rn).
Personally will try to find some time to work on SwiftGen a little next week-end, but there are other breaking PRs pending, so won't probably be enough to reach a release of 6.0 which needs those breaking PRs to be merged first… We'll do our best.

**Comment by djbe** [comment]
@vkaramov SwiftGen 6.0.0 has been released (we're even at 6.0.2), which should address your issue. I'm closing this for now, but you can re-open it if the issue is still there.
